fl4health/strategies/feddg_ga_strategy.py

```diff
@@ -0,0 +1,382 @@
+
+    def update_weights_by_ga(self, server_round: int, cids: List[str]) -> None:
+        """
+        Update the self.adjustment_weights dictionary by calculating the new weights
+        based on the current server round, fit and evaluation metrics.
+
+        Args:
+            server_round: (int) the current server round.
+            cids: (List[str]) the list of client ids that participated in this round.
+        """
+        value_list = []
+        # calculating local vs global metric difference
+        for cid in cids:
+            assert (
+                cid in self.train_metrics and cid in self.evaluation_metrics
+            ), f"{cid} not in {self.train_metrics.keys()} or {self.evaluation_metrics.keys()}"
+
+            global_model_metric_value = self.evaluation_metrics[cid][self.fairness_metric.metric_name]
+            local_model_metric_value = self.train_metrics[cid][self.fairness_metric.metric_name]
+            assert isinstance(global_model_metric_value, float) and isinstance(local_model_metric_value, float)
+
+            value_list.append(global_model_metric_value - local_model_metric_value)
+
+        # calculating norm gap
+        value_list_ndarray = np.array(value_list)
+        max_value = np.max(np.abs(value_list_ndarray))
+
+        if max_value == 0:
+            log(
+                WARNING,
+                "Max value in metric diff list is 0. Adjustment weights will remain the same. "
+                + f"Value list: {value_list}",
+            )
+            norm_gap_list = np.zeros(len(value_list))
+        else:
+            norm_gap_list = value_list_ndarray / max_value
+
+        step_size = (1.0 / 3.0) * self.get_current_weight_step_size(server_round)
+
+        # updating weights
+        new_total_weight = 0.0
+        for i in range(len(cids)):
+            cid = cids[i]
+            self.adjustment_weights[cid] += self.fairness_metric.signal * norm_gap_list[i] * step_size
+
+            # weight clip
+            clipped_weight = np.clip(self.adjustment_weights[cid], 0.0, 1.0)
+            self.adjustment_weights[cid] = clipped_weight
+            # TODO Question: should we sum all the weights or just the current clients' weights?
+            new_total_weight += clipped_weight
+
+        for cid in cids:
+            self.adjustment_weights[cid] /= new_total_weight
+
+    def get_current_weight_step_size(self, server_round: int) -> float:
+        """
+        Calculates the current weight step size based on the current server round,  weight
+        step size and total number of rounds.
+
+        Args:
+            server_round: (int) the current server round
+
+        Returns: (float) the current value for the weight step size.
+        """
+        assert self.num_rounds is not None
+        weight_step_size_decay = self.weight_step_size / self.num_rounds
+        weight_step_size_for_round = self.weight_step_size - ((server_round - 1) * weight_step_size_decay)
```

Comment:
I agree with Fatemeh, I think the way this is written my not be algebraically equivalent actually, if I'm reading the variables correctly.

Reply:
Again, pulled from the implementation:

https://github.com/MediaBrain-SJTU/FedDG-GA/blob/master/algorithms/fedavg/train_pacs_GA.py#L83

I guess next time I should rely more on the paper instead  😅

Reply:
So I'm honestly not sure what they are doing here, as their round index starts at 0. So subtracting 1 from it makes no sense. However, in the paper, their first local round of training starts at zero. So for them $(1-r/R)\cdot d=d$ for the first local round. If we start at `server_round = 1` (which I think we do), then doing $d\cdot (1-(r-1)/R)$ as you have coded gets us back to that expression correctly. 
1.  So if you confirm that our server rounds start at 1, then your formulation is correct and we should add a comment that this differs from the paper because our rounds start at 1 not 0.
2. If our server rounds start at 0 (I don't think so but not 100% sure), then we need to use the paper formulation
3. There implementation is dumb either way 😂

Reply:
Our server round starts at 1. I'm confused now. In our case, I should keep this code as is?

Reply:
makes sense, thanks!!!